File: teacher/process_excel.py
#!/usr/bin/env python3
"""
Excel Comment Processor using Pandas/OpenPyXL
Processes Excel files with multiple sheets, calculates scores from column L,
and adds comments to column M based on configurable rules.

Author: AI Assistant
Date: 2024
"""

import sys
import json
import os
import pandas as pd
import openpyxl
from openpyxl.utils import get_column_letter
from pathlib import Path

def log_debug(message):
    """Log debug message to file"""
    log_file = Path(__file__).parent / '..' / 'logs' / 'excel_processor.log'
    log_file.parent.mkdir(exist_ok=True)

    from datetime import datetime
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    with open(log_file, 'a', encoding='utf-8') as f:
        f.write(f"[{timestamp}] {message}\n")

def get_comment(score, rules):
    """Get comment based on score and rules"""
    try:
        score = float(score)
        for rule in rules:
            if rule['min'] <= score <= rule['max']:
                return rule['comment']
        return ''
    except (ValueError, TypeError):
        return ''

def process_excel_with_pandas(input_file, output_file, comment_rules):
    """
    Process Excel file using Pandas and OpenPyXL
    """
    try:
        log_debug(f"Processing Excel file: {input_file}")

        # Handle temporary files by copying to a proper .xlsx extension if needed
        import tempfile
        import shutil

        temp_input_file = input_file
        if not input_file.lower().endswith(('.xlsx', '.xlsm', '.xltx', '.xltm')):
            # Create a temporary file with .xlsx extension
            with tempfile.NamedTemporaryFile(suffix='.xlsx', delete=False) as temp_file:
                temp_input_file = temp_file.name
            shutil.copy2(input_file, temp_input_file)
            log_debug(f"Copied temp file to: {temp_input_file}")

        # Load the Excel file with openpyxl to preserve formatting
        wb = openpyxl.load_workbook(temp_input_file, data_only=True)

        # Process each sheet
        for sheet_name in wb.sheetnames:
            ws = wb[sheet_name]
            log_debug(f"Processing sheet: {sheet_name}")

            # Find the last row with data
            max_row = ws.max_row

            # Start from row 8 (1-based indexing)
            start_row = 8

            for row in range(start_row, max_row + 1):
                try:
                    # Get STT from column A (1-based)
                    stt_cell = ws.cell(row=row, column=1)
                    stt = str(stt_cell.value).strip() if stt_cell.value is not None else ''

                    # Get student code from column B
                    code_cell = ws.cell(row=row, column=2)
                    student_code = str(code_cell.value).strip() if code_cell.value is not None else ''

                    # Skip if STT is empty, not numeric, or student code is empty
                    if not stt or not stt.isdigit() or not student_code:
                        continue

                    # Get score from column L (12th column, 1-based)
                    score_cell = ws.cell(row=row, column=12)
                    score_value = score_cell.value

                    # Handle different score formats
                    if score_value is None:
                        score = 0.0
                    elif isinstance(score_value, (int, float)):
                        score = float(score_value)
                    else:
                        # Try to parse string values
                        try:
                            score = float(str(score_value).replace(',', '.'))
                        except (ValueError, AttributeError):
                            score = 0.0

                    log_debug(f"Row {row}: STT={stt}, Code={student_code}, Score={score}")

                    # Get comment
                    comment = get_comment(score, comment_rules)

                    # Write comment to column M (13th column, 1-based)
                    comment_cell = ws.cell(row=row, column=13)
                    comment_cell.value = comment

                    log_debug(f"Added comment: {comment}")

                except Exception as e:
                    log_debug(f"Error processing row {row}: {str(e)}")
                    continue

        # Save the workbook with optimized settings
        wb.save(output_file)
        log_debug(f"Saved processed file: {output_file}")

        return True

    except Exception as e:
        log_debug(f"Error processing Excel file: {str(e)}")
        return False

def main():
    """Main processing function"""
    try:
        # Read input from stdin (JSON)
        input_data = json.load(sys.stdin)

        input_file = input_data['input_file']
        output_file = input_data['output_file']
        comment_rules = input_data['comment_rules']

        log_debug(f"Starting processing: {input_file} -> {output_file}")

        # Process the file
        success = process_excel_with_pandas(input_file, output_file, comment_rules)

        if success:
            file_size = os.path.getsize(output_file)
            result = {
                'success': True,
                'file_size': round(file_size / 1024 / 1024, 2),
                'message': 'File processed successfully with Pandas/OpenPyXL'
            }
        else:
            result = {
                'success': False,
                'message': 'Failed to process file with Pandas/OpenPyXL'
            }

        # Output result as JSON
        print(json.dumps(result))

    except Exception as e:
        log_debug(f"Main error: {str(e)}")
        result = {
            'success': False,
            'message': f'Error: {str(e)}'
        }
        print(json.dumps(result))

if __name__ == '__main__':
    main()

File: teacher/process_excel.php
<?php
include '../includes/session_check.php'; // Ensure logged in

// Check if teacher (not admin)
if (!isset($_SESSION['username']) || $_SESSION['username'] === 'admin') {
    header('Location: ../login.php');
    exit;
}

// Increase memory limit for large Excel files
ini_set('memory_limit', '2048M');
ini_set('max_execution_time', 300); // 5 minutes

// Cấu hình mốc điểm và nhận xét - load from localStorage or use defaults
$default_rules = [
    ['min' => 0,   'max' => 3.5, 'comment' => 'Chưa đạt, ý thức học kém, cần cố gắng nhiều hơn.'],
    ['min' => 3.5, 'max' => 5.0, 'comment' => 'Chưa đạt, chưa cố gắng học, cần nghiêm túc hơn.'],
    ['min' => 5.0, 'max' => 6.5, 'comment' => 'Có cố gắng, cần phát huy thêm.'],
    ['min' => 6.5, 'max' => 8.0, 'comment' => 'Siêng học, cần phát huy thêm.'],
    ['min' => 8.0, 'max' => 10.0, 'comment' => 'Chăm chỉ học tập, rất tích cực phát biểu, gương mẫu cho học sinh.']
];

// Try to load custom rules from POST data or use defaults
$comment_rules = isset($_POST['comment_rules']) ? json_decode($_POST['comment_rules'], true) : $default_rules;
if (!$comment_rules) {
    $comment_rules = $default_rules;
}

// Hàm ghi log debug
function logDebug($message) {
    $logFile = __DIR__ . '/../logs/debug.log';
    $timestamp = date('Y-m-d H:i:s');
    file_put_contents($logFile, "[$timestamp] $message\n", FILE_APPEND);
}

// Hàm xử lý Excel bằng Python/LibreOffice
function processExcelWithPython($inputFile, $outputFile, $commentRules) {
    $pythonScript = __DIR__ . '/process_excel.py';

    // Kiểm tra Python script tồn tại
    if (!file_exists($pythonScript)) {
        return [
            'success' => false,
            'message' => 'Python processing script not found'
        ];
    }

    // Tạo dữ liệu input cho Python script
    $inputData = [
        'input_file' => $inputFile,
        'output_file' => $outputFile,
        'comment_rules' => $commentRules
    ];

    // Chạy Python script
    $cmd = "\"C:\\Python313\\python.exe\" \"$pythonScript\"";
    $descriptors = [
        0 => ['pipe', 'r'], // stdin
        1 => ['pipe', 'w'], // stdout
        2 => ['pipe', 'w']  // stderr
    ];

    $process = proc_open($cmd, $descriptors, $pipes);

    if (!is_resource($process)) {
        return [
            'success' => false,
            'message' => 'Failed to start Python process'
        ];
    }

    // Gửi dữ liệu input
    fwrite($pipes[0], json_encode($inputData));
    fclose($pipes[0]);

    // Đọc output
    $output = stream_get_contents($pipes[1]);
    $error = stream_get_contents($pipes[2]);

    fclose($pipes[1]);
    fclose($pipes[2]);

    $returnCode = proc_close($process);

    logDebug("Python script output: $output");
    if (!empty($error)) {
        logDebug("Python script error: $error");
    }
    logDebug("Python script return code: $returnCode");

    // Parse JSON output
    $result = json_decode($output, true);
    if (json_last_error() !== JSON_ERROR_NONE) {
        return [
            'success' => false,
            'message' => 'Invalid JSON output from Python script: ' . $output
        ];
    }

    return $result;
}

if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_FILES['excelFile'])) {
    $file = $_FILES['excelFile'];

    // Kiểm tra lỗi upload
    if ($file['error'] !== UPLOAD_ERR_OK) {
        header('Content-Type: application/json');
        echo json_encode([
            'success' => false,
            'message' => 'Lỗi upload file: ' . $file['error']
        ]);
        exit;
    }

    // Kiểm tra định dạng file
    $fileExtension = strtolower(pathinfo($file['name'], PATHINFO_EXTENSION));
    if ($fileExtension !== 'xlsx') {
        header('Content-Type: application/json');
        echo json_encode([
            'success' => false,
            'message' => 'Chỉ chấp nhận file .xlsx'
        ]);
        exit;
    }

    try {
        // Create temp directory for processing
        $tempDir = sys_get_temp_dir() . '/cvd_excel_processing/';
        if (!is_dir($tempDir)) {
            mkdir($tempDir, 0755, true);
        }

        $inputFile = $file['tmp_name'];
        $originalName = $file['name'];
        $outputFile = $tempDir . 'processed_' . time() . '_' . basename($originalName);

        logDebug("Starting Excel processing: $inputFile -> $outputFile");

        // Xử lý file bằng Python/LibreOffice
        $result = processExcelWithPython($inputFile, $outputFile, $comment_rules);

        if ($result['success']) {
            // Get file size for logging
            $fileSize = filesize($outputFile);
            logDebug("Processed file size: " . round($fileSize / 1024 / 1024, 2) . " MB");

            // Return JSON response with download URL
            header('Content-Type: application/json');
            echo json_encode([
                'success' => true,
                'downloadUrl' => 'download_processed.php?file=' . urlencode(basename($outputFile)),
                'fileSize' => round($fileSize / 1024 / 1024, 2) . ' MB',
                'message' => 'File processed successfully with LibreOffice/Python!'
            ]);
        } else {
            header('Content-Type: application/json');
            echo json_encode([
                'success' => false,
                'message' => 'Error processing file: ' . ($result['message'] ?? 'Unknown error')
            ]);
        }

    } catch (Exception $e) {
        logDebug("Error: " . $e->getMessage());
        header('Content-Type: application/json');
        echo json_encode([
            'success' => false,
            'message' => 'Error processing file: ' . $e->getMessage()
        ]);
    }
} else {
    header('Content-Type: application/json');
    echo json_encode([
        'success' => false,
        'message' => 'Invalid request'
    ]);
}
?>
